### PR TITLE
[cuda] Move to hal/drivers and wire up BUILD files

### PR DESCRIPTION
**Pull request by antiagainst**
This commit moves the CUDA HAL driver rewrite to the
`hal/drivers` directory given it's functional and ready for
normal usage. By this we can start run tests with CI
to make sure it does not regress. Further improvements
can happen directly in this directory.

This provides an easy route for trying out the rewrite before
eventually replace the existing HAL driver. 

Along the way wired up BUILD configurations.

Progress towards https://github.com/openxla/iree/issues/13245

### CODE REVIEW

File: runtime/src/iree/hal/drivers/cuda2/cuda_device.h [hunk 1]
```diff
@@ -7,11 +7,11 @@
 #ifndef EXPERIMENTAL_CUDA2_CUDA_DEVICE_H_
 #define EXPERIMENTAL_CUDA2_CUDA_DEVICE_H_
```
Comment (benvanik) [inline runtime/src/iree/hal/drivers/cuda2/cuda_device.h]: header guards need updating

**Review by benvanik — CHANGES_REQUESTED** [review]

File: CMakeLists.txt [hunk 2]
```diff
@@ -226,6 +226,7 @@ option(IREE_HAL_DRIVER_DEFAULTS "Sets the default value for all runtime HAL driv
 set(IREE_HAL_DRIVER_CUDA_DEFAULT ${IREE_HAL_DRIVER_DEFAULTS})
 if(NOT IREE_CUDA_AVAILABLE OR CMAKE_CROSSCOMPILING)
   set(IREE_HAL_DRIVER_CUDA_DEFAULT OFF)
+  set(IREE_HAL_DRIVER_CUDA2_DEFAULT OFF)
```
Comment (ScottTodd) [inline CMakeLists.txt]: Initialize to `ON` if available? Should be explicit either way.
```suggestion
set(IREE_HAL_DRIVER_CUDA_DEFAULT ${IREE_HAL_DRIVER_DEFAULTS})
set(IREE_HAL_DRIVER_CUDA2_DEFAULT ${IREE_HAL_DRIVER_DEFAULTS})
if(NOT IREE_CUDA_AVAILABLE OR CMAKE_CROSSCOMPILING)
  set(IREE_HAL_DRIVER_CUDA_DEFAULT OFF)
  set(IREE_HAL_DRIVER_CUDA2_DEFAULT OFF)
```

File: runtime/src/iree/hal/drivers/cuda2/README.md
Comment (ScottTodd) [inline runtime/src/iree/hal/drivers/cuda2/README.md:1]: Could tweak the README while moving it. In particular this section: https://github.com/openxla/iree/blob/4a347236e7a9ab38b4dffe9e089e72e15cf3822c/experimental/cuda2/README.md?plain=1#L6-L10

(also note that the issue link there does not render on GitHub, see https://github.com/openxla/iree/blob/main/experimental/cuda2/README.md)

File: runtime/src/iree/hal/drivers/cuda2/cts/CMakeLists.txt [hunk 3]
```diff
@@ -8,15 +8,15 @@ iree_hal_cts_test_suite(
...
   COMPILER_TARGET_BACKEND
     "cuda"
   EXECUTABLE_FORMAT
     "\"PTXE\""
   DEPS
-    iree::experimental::cuda2::registration
+    iree::hal::drivers::cuda2::registration
   EXCLUDED_TESTS
     # HAL event is unimplemented for now.
     "event"
```
Comment (ScottTodd) [inline runtime/src/iree/hal/drivers/cuda2/cts/CMakeLists.txt:51]: Are events still unimplemented?

File: CMakeLists.txt (hunk 2 above)
Comment (antiagainst) [reply CMakeLists.txt]: Good catch. Missed that. Fixed.

File: runtime/src/iree/hal/drivers/cuda2/cts/CMakeLists.txt (hunk 3 above)
Comment (antiagainst) [reply runtime/src/iree/hal/drivers/cuda2/cts/CMakeLists.txt:51]: Right. The compiler side does not generate this at the moment; so it's fine.

File: runtime/src/iree/hal/drivers/cuda2/README.md
Comment (antiagainst) [reply runtime/src/iree/hal/drivers/cuda2/README.md:1]: Revised.

File: CMakeLists.txt [hunk 4]
```diff
@@ -243,6 +245,7 @@ if(NOT APPLE OR NOT ${CMAKE_SYSTEM_PROCESSOR} MATCHES "arm64")
 endif()
 
 option(IREE_HAL_DRIVER_CUDA "Enables the 'cuda' runtime HAL driver" ${IREE_HAL_DRIVER_CUDA_DEFAULT})
+option(IREE_HAL_DRIVER_CUDA2 "Enables the 'cuda2' runtime HAL driver" ${IREE_HAL_DRIVER_CUDA2_DEFAULT})
```
Comment (ScottTodd) [inline CMakeLists.txt:263]: Can we come up with a plan for switching from the old `cuda` to the new `cuda2`? Maybe post an RFC and an update to https://github.com/openxla/iree/issues/13245, and could link to one or both of those here.

Some ideas:

* Drive to feature complete
  * All CTS tests and execution tests we care about pass using `cuda2`
  * Figure out the story for flags like on https://github.com/openxla/iree/issues/13984
* Switch default in-repo
  * All in-repo benchmarks using `cuda2`, without egregious regressions
* Delete (old) `cuda` and rename `cuda2` to `cuda`

Ideally we'd have a timeline in mind so we don't stay in this state of having two similar paths for long.

**Review by ScottTodd — APPROVED** [review]

File: CMakeLists.txt (hunk 4 above)
Comment (antiagainst) [reply CMakeLists.txt:263]: SG. I've commented on #13245 with https://github.com/openxla/iree/issues/13245#issuecomment-1684210183. I'll send out a mail later too.

Comment (antiagainst) [reply CMakeLists.txt:263]: Hard for me to give a precise timeline though; given I'm multiplexing over quite a few work streams. But yeah agreed we don't want to be in a hybrid state for too long; I'm hoping to get feature parity down soon.

**Comment by antiagainst** [comment]
Hi @benvanik, I have addressed all known issues so this is ready for another review now. After moving I'd be switching the tests/benchmarks to use the new impl to test out.

**Comment by benvanik** [comment]
woo excited to get rid of the old impl!

**Review by benvanik — APPROVED** [review]

**Pull request merged**